syft/pkg/cataloger/golang/version.go

```diff
@@ -1,226 +1,99 @@
-func isExe(file string, info fs.FileInfo) bool {
+func isExe(file string, mode os.FileMode) bool {
 	if runtime.GOOS == "windows" {
 		return strings.HasSuffix(strings.ToLower(file), ".exe")
 	}
-	return info.Mode().IsRegular() && info.Mode()&0111 != 0
+	return mode.IsRegular() && mode&0111 != 0
 }
 
 // scanFile scans file to try to report the Go and module versions.
-// If mustPrint is true, scanFile will report any error reading file.
-// Otherwise (mustPrint is false, because scanFile is being called
-// by scanDir) scanFile prints nothing for non-Go executables.
-func scanFile(file string, info fs.FileInfo, mustPrint bool) {
-	if info.Mode()&fs.ModeSymlink != 0 {
-		// Accept file symlinks only.
-		i, err := os.Stat(file)
-		if err != nil || !i.Mode().IsRegular() {
-			if mustPrint {
-				fmt.Fprintf(os.Stderr, "%s: symlink\n", file)
-			}
-			return
-		}
-		info = i
-	}
-
-	if !isExe(file, info) {
-		if mustPrint {
-			fmt.Fprintf(os.Stderr, "%s: not executable file\n", file)
-		}
-		return
+func scanFile(reader unionReader, filename string, mode os.FileMode) []*debug.BuildInfo {
+	if !isExe(filename, mode) {
+		log.Debugf("golang cataloger: %s: not executable file\n", filename)
+		return nil
 	}
 
-	x, err := openExe(file)
+	// NOTE: multiple readers are returned to cover universal binaries, which are files
+	// with more than one binary
+	readers, err := getReaders(reader)
 	if err != nil {
-		if mustPrint {
-			fmt.Fprintf(os.Stderr, "%s: %v\n", file, err)
-		}
-		return
+		log.Warnf("golang cataloger: opening binary: %v", err)
+		return nil
 	}
-	defer x.Close()
 
-	vers, mod := findVers(x)
-	if vers == "" {
-		if mustPrint {
-			fmt.Fprintf(os.Stderr, "%s: go version not found\n", file)
+	builds := make([]*debug.BuildInfo, 0)
+	for _, r := range readers {
+		bi, err := buildinfo.Read(r)
+		if err != nil {
+			if pathErr := (*os.PathError)(nil); !errors.As(err, &pathErr) {
+				log.Warnf("golang cataloger: scanning file %s: %v\n", filename, err)
+			}
+			return nil
 		}
-		return
+		builds = append(builds, bi)
 	}
 
-	fmt.Printf("%s: %s\n", file, vers)
-	if *versionM && mod != "" {
-		fmt.Printf("\t%s\n", strings.ReplaceAll(mod[:len(mod)-1], "\n", "\n\t"))
-	}
+	setArch(readers, builds)
+
+	return builds
 }
-*/
-
-// The build info blob left by the linker is identified by
-// a 16-byte header, consisting of buildInfoMagic (14 bytes),
-// the binary's pointer size (1 byte),
-// and whether the binary is big endian (1 byte).
-var buildInfoMagic = []byte("\xff Go buildinf:")
-
-// findVers finds and returns the Go version and module version information
-// in the executable x.
-func findVers(x exe) (vers, mod string) {
-	// Read the first 64kB of text to find the build info blob.
-	text := x.DataStart()
-	data, err := x.ReadData(text, 64*1024)
+
+// openExe opens file and returns it as io.ReaderAt.
+func getReaders(f unionReader) ([]io.ReaderAt, error) {
+	data := make([]byte, 16)
```

Comment:
the 16-byte slice + readfull + seek seems like dead code -- is this doing anything?

Reply:
you're right, removing it